quickwit/quickwit-search/src/find_trace_ids_collector.rs

```diff
@@ -0,0 +1,508 @@
+// Copyright (C) 2023 Quickwit, Inc.
+//
+// Quickwit is offered under the AGPL v3.0 and as commercial software.
+// For commercial licensing, contact us at hello@quickwit.io.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::cmp::{Ord, Ordering};
+use std::collections::{BinaryHeap, HashSet};
+use std::hash::{Hash, Hasher};
+use std::sync::Arc;
+
+use base64::prelude::{Engine, BASE64_STANDARD};
+use base64::DecodeSliceError;
+use fnv::{FnvHashMap, FnvHashSet};
+use itertools::Itertools;
+use quickwit_common::binary_heap;
+use serde::{Deserialize, Serialize};
+use tantivy::collector::{Collector, SegmentCollector};
+use tantivy::fastfield::{Column, MultiValuedFastFieldReader};
+use tantivy::{DateTime, DocId, InvertedIndexReader, Score, SegmentReader};
+
+type B64TraceId = [u8; 24];
+type TermOrd = u64;
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(into = "TraceIdFruitForSerde")]
+#[serde(try_from = "TraceIdFruitForSerde")]
+pub struct TraceIdFruit {
+    pub trace_id: B64TraceId,
+    pub span_timestamp: DateTime,
+}
+
+impl TraceIdFruit {
+    pub fn new(trace_id_bytes: &[u8], span_timestamp: DateTime) -> Self {
```

Comment:
There is a hidden contract here. Trace_id_bytes needs to have a specific len otherwise it panics, and this struct and method are pub.
Ideally we should get rid of the hidden contract.

Reply:
Or at least, we  want to make that private.

Reply:
The size of the bytes is the result of the index containing the right information. This is too loose. I think we need to return an error if trace_id_bytes does not have the right size.

Reply:
Changed to `&[u8; 24]`.